### ISSUE TITLE
Node v18 Released

### ISSUE THREAD

**Issue by chenrui333** [aws/aws-lambda-base-images#47]
Even though we are still waiting for node v16 base image, this is the issue tracker for node v18 base image

https://nodejs.org/en/blog/announcements/v18-release-announce/

**Comment by jlarmstrongiv** [comment]
I’ve been waiting for someone to do this 🤣 thanks

**Comment by luizcorreia** [comment]
Maybe that way, the Node v18 will be done before past one year.

**Comment by mikepage** [comment]
Best thing you could do to enjoy Node v18 like-features is move to Netlify. They released Edge functions with Deno deploy: https://www.netlify.com/blog/announcing-serverless-compute-with-edge-functions 

Do tell Vercel, they may reconsider using AWS lambda images :)

**Comment by chenrui333** [comment]
I honestly mind receiving flood of email notifications, we really need to get more ppl to pressure AWS stay on top of this nodejs runtime update game.

**Comment by vith** [comment]
> Honestly, having a 6 month bake (releasing the april LTS release in around October for lambda) would be a good compromise. gives enough time for tests and means we're only "one version behind"

Isn't that what the 6 months between v18 release on 2022-04-19 and v18 becoming LTS on 2022-10-25 is for?

Amazon can start making v18 images available *now* if they want to have 6 months of testing before offering v18 as LTS.

**Comment by dl748** [comment]
Should be noted as well, that they will be transitioning to Javascript SDK V3 as well for the default SDK.

**Comment by adcreare** [comment]
Like many others I agree with the sentiments, AWS needs to go better on this. 
I suggest everyone bug their account managers, technical account managers and log support tickets if you haven’t already. 

I went back and dug out the release history of lambda back to node 8. It is not a pretty sight.

The current delay is second only to the delay with the node 10 release that took until May 13.  After Node 10 things got better - Node 12 dropped very promptly, 14 it started to drift back out and with 16 we’re back to at least May. 

Interesting to note the other two major cloud providers, ie Google and Azure, did manage to get 16 out in a timely manner. 

```
Node 16: 
	Node LTS: 2021-10-26
	AWS Lambda Release:  ???
	Google Cloud Functions release: 2021-11-17
	Azure  Functions release: 2021-11-22
Node 14: 
	Node LTS: 2020-10-20
	AWS Lambda Release: 2021-02-03
Node 12: 
	Node LTS: 2019-10-22 
	AWS Lambda Release: 2019-11-18
Node 10:
	Node LTS: 2018-10-30
	AWS Lambda Release: 2019-05-13
Node 8:
	Node LTS: 2017-10-31
	AWS Lambda Release: 2018-04-02
```

**Comment by paul-uz** [comment]
@adcreare I mean, it's not like Amazon is a multi-billion dollar company.... oh wait

**Comment by jtuliani** [comment]
Thanks all for the keen interest in seeing Node releases land promptly in Lambda.

Regarding Node releases in general, we're tracking the Active LTS release dates in October, not the 'Current' release dates in April. This is because, as stated at https://nodejs.org/en/about/releases/, "Production applications should only use Active LTS or Maintenance LTS releases".

As @adcreare pointed out above, there have been delays between Active LTS and our corresponding GA releases, not least for Node 16 (see #14 for latest status). While I can't give an ETA for Node 18 yet, we do hear your feedback and this is something we're actively working to improve.

**Comment by faradaytrs** [comment]
@jtuliani Well you could add it as beta status or something, actually node 12 is also not usable in production because it's not supported anymore, but you still have this runtime supported. I think it's nothing bad about thinking of new releases in advance. "Current" version is stable enough for some simple applications. It's a good idea to release it right when it hits lts status.

**Comment by GrahamCampbell** [comment]
> Regarding Node releases in general, we're tracking the Active LTS release dates in October, not the 'Current' release dates in April. This is because, as stated at https://nodejs.org/en/about/releases/, "Production applications should only use Active LTS or Maintenance LTS releases".

Then why is your node 16 image that is just released based on node 16.4, which by your version definition, should not be used in production, because it was released multiple months before the first 16.x LTS version!!!

**Comment by dl748** [comment]
> @jtuliani Well you could add it as beta status or something

This is "close" to what I'd like. i'm thinking more like `nodejs18.x-experimental` that can only to assigned through the API/CLI and not through the console (currently how 16 is working now). Where, if they are are worried about support, have a flag that needs to be sent via API, `"agree_experimental": true` when sending a runtime of "*-experimental", where the usage agreement is "use at your own risk"....  then when GA is ready or close, they make a `nodejs18.x`

This would allow us to deploy an experimental version and allow us to fix bugs ahead of the GA releases.

If someone wants to use this for production purposes, that's their issue then as they agreed to the terms.

**Comment by dl748** [comment]
There shouldn't be a need to remove any configs, however it would be more aggressive on the removal of the runtimes. I would just say, create a new GA runtime without the suffix, and either remove the *-experimental runtime immediately or after a month. They already have policy and protocols for deprecation. The only other thing is, I'd probably disallow the runtime's for @Edge.

The nice thing about it is. Once a GA is announced then they deprecate and immediately release the "next" LTS version. So, if they have node18.x-experimental... after the GA, they release node18.x runtime, where both runtimes exist, then they release node20.x-experimental and remove node18.x-experimental. Experimentals are just treated as another runtime.

**Comment by GrahamCampbell** [comment]
> but would have naturally transitioned at GA without any actions on the part of the customer

However that's not true, because many major bugs and security issues were fixed since the 16.4 release.

**Comment by jtuliani** [comment]
@GrahamCampbell Please can you clarify how you checked the Node 16 version and which region you used? When I tried a simple function to `console.log(process.version);` just now in `eu-west-1`, I see the Node version reported as `v16.14.0`. Thanks!

**Comment by dl748** [comment]
@barneyparker The same way they handle deprecated versions of runtimes now. https://docs.aws.amazon.com/lambda/latest/dg/lambda-runtimes.html#runtime-support-policy
According to documentation
```
Phase 1 - Lambda no longer applies security patches or other updates to the runtime. You can no longer create functions that use the runtime, but you can continue to update existing functions.

Phase 2 - you can no longer create or update functions that use the runtime. 

Lambda does not block invocations of functions that use deprecated runtime versions. Function invocations continue indefinitely after the runtime version reaches end of support.
```

So I would expect, as a completely separate runtime, it would follow these steps but at a "faster" pace, and there would only be one "experimental" per GA release

**Comment by dl748** [comment]
@jtuliani The Main reason is for developers to deploy a "test" version of their application.  This would allow us to iron out any issues with a new version before (90-99%) a GA is deployed. As it is now, when a GA is released, it can be up to a month sometimes for testing and QA teams to work with before we can utilize this in a production environment. Then all that is left is any issues between a "preview" version and the GA runtimes, which should be low.

**Comment by willfarrell** [comment]
As the core maintainer of [Middy](https://github.com/middyjs/middy), being able to know what will be included in the runtime when it hits GA will allow myself and other OSS projects to better server the AWS Lambda community. Currently, we need to be over conservative or reactionary with major releases which I feel isn't serving the community as well as it could be. Also, having more lead time mean better planning and less rushing to test and publish a new version once GA is released. Frameworks need to be updated and released before, as @dl748 mentioned, other downstream projects/developers can use them.

I like this idea of having the LTS runtime released via to the API/SDK with a simple changelog in GitHub to know if there are any other changes besides the runtime itself. Then having a switch to GA it once the version transitions to LTS, enabling the console and updating the public docs. I don't think the runtime needs a `-experimental` suffix, just remove the API flag when no longer needed.

Having access to the latest runtime sooner has additional advantages:
- faster, which lowers costs and can be a competitive edge
- fewer polyfills needed (ie fetch, web stream, etc)
- removal of experimental flags (ie json import)
- simpler and safer build process, js may no longer need to be transpiled

Happy to discuss feedback on preview nodejs runtimes and feature requests I've been collecting over the years. The Middy team is already building a relationship with the [AWS Lambda Powertools Typescript](https://github.com/awslabs/aws-lambda-powertools-typescript) team to see how we can better work together to serve the community, would love to do the same with the Lambda nodejs runtime team if there is interest.

cc @lmammino @saragerion @dreamorosi @ijemmy @flochaz

**Comment by mbklein** [comment]
@jtuliani Could there maybe be a way to differentiate pre-released runtimes from released ones? Something like the `CAPABILITY` flags you have to include in some CloudFormation calls to confirm that you know you're doing what you think you're doing? For example, if I tried to execute `CreateFunction` with a runtime of `nodejs18.x` but without some kind of `IncludePrereleaseRuntimes` parameter, I'd get an `InvalidParameterValueException` (or something). Including that flag lets me select unsupported runtimes, with the full knowledge that what I'm doing isn't (yet) supported. I think it'd provide a good balance of trusting developers/users with the flexibility to do what they want while still making sure you're not fielding support requests for things in pre-release state.

That said, I understand that changing the public API is way harder than some of the other solutions presented. One way around that might be to change the actual name of the runtime to, say, `nodejs18.x-pre` until it's supported. Anything that makes selecting it a deliberate act that forces acknowledgment of its prerelease status.

**Comment by paulgalow** [comment]
I like the idea of implementing Node 18 current with a runtime name of `nodejs18.x`. I see a chance this might make it easier for adjacent software projects to keep their runtime support up to date and implement support by the time Node 18 LTS is released. 

For example: Frameworks like [Serverless Framework](https://github.com/serverless/serverless/pull/11049), plugins (like [serverless-esbuild](https://github.com/floydspace/serverless-esbuild/pull/306)) or IDE extensions like [Serverless IDE](https://github.com/threadheap/serverless-ide-vscode/pull/163). All of them have/had to add support for the latest runtime and all of them have been blocked/waiting until May 9 to get started, making an already late runtime release take even longer because of the time required for dependencies to follow suit.

If AWS released `nodejs18.x` before Node 18 LTS, all of those projects could start working on support for it months earlier. And as mentioned before, we all could run and test Node 18 current much earlier to iron out issues. So by the time LTS hits, we could have a much more mature state of support throughout the ecosystem.

**Comment by pbadenski** [comment]
FYI As AWS is painfully slow at releasing new versions of node images we build our own lambda layers. At the moment we are not even able to do that as Node 18 requires glibc >= 2.28 currently unsupported by Amazon Linux (1 or 2). What I think it means - is that we will not have Node 18 on AWS Lambda until it's operating on Amazon Linux 2022.

**Comment by willfarrell** [comment]
NodeJS just moved up the EoL for v16 to 2023-09-11! The race is now on for all vendors to release v18 ASAP to ensure customers have sufficient time to upgrade. Hopefully this will force a new release schedule.

https://nodejs.org/en/blog/announcements/nodejs16-eol

**Comment by lmammino** [comment]
I totally agree with @willfarrell that this is becoming much more of a priority right now. Hopefully AWS can prioritise this work

**Comment by jlarmstrongiv** [comment]
> NodeJS just moved up the EoL for v16 to 2023-09-11! The race is now on for all vendors to release v18 ASAP to ensure customers have sufficient time to upgrade. Hopefully this will force a new release schedule.

Probably a good time to migrate from AWS Linux 2 base image to the AWS Linux 2022 base image too

**Comment by pbadenski** [comment]
Analysis of 5 days of data shows cold start durations: [min: 1000 ms, avg: 1500 ms, max: 3500 ms]. These can be quite significant, however - we're not that sensitive to cold starts and only 0.5% (1 in 200) of our requests are cold starts.

Hope this helps.

P. S. Thank you for your question - cold starts never been an issue for us, so I've never checked it. I've learnt something! :)

**Comment by paul-uz** [comment]
> Analysis of 5 days of data shows cold start durations: [min: 1000 ms, avg: 1500 ms, max: 3500 ms]. These can be quite significant, however - we're not that sensitive to cold starts and only 0.5% (1 in 200) of our requests are cold starts.
> 
> Hope this helps.
> 
> P. S. Thank you for your question - cold starts never been an issue for us, so I've never checked it. I've learnt something! :)

Could you tell me how to best gather metrics on cold (and hot) start times?

**Comment by hutchy2570** [comment]
> Analysis of 5 days of data shows cold start durations: [min: 1000 ms, avg: 1500 ms, max: 3500 ms]. These can be quite significant, however - we're not that sensitive to cold starts and only 0.5% (1 in 200) of our requests are cold starts.
> 
> Hope this helps.
> 
> P. S. Thank you for your question - cold starts never been an issue for us, so I've never checked it. I've learnt something! :)

What is your memory configuration for these metrics?

**Comment by pbadenski** [comment]
@paul-uz https://georgemao.medium.com/analyze-your-aws-lambda-logs-with-cloudwatch-logs-insights-170a5b58225a

@hutchy2570 2048MB

**Comment by fromtheexchange** [comment]
@jtuliani following up from https://github.com/aws/aws-lambda-base-images/issues/14#issuecomment-1120864028

> @fromtheexchange That's a great question regarding the alignment of RICs. This is something we currently plan to address in Node 18.

In addition, the official `amazon/aws-lambda-nodejs` image on https://hub.docker.com/r/amazon/aws-lambda-nodejs needs to be updated to include the `linux/arm64` architecture.

Finally, it may be worth looking into the base image migration to amazonlinux:2022 from amazonlinux:2

Thank you for improving NodeJS support in AWS Lambda 🚀

**Comment by jtuliani** [comment]
@fromtheexchange Thank you for the suggestions. We are aware of the AL2 -> AL2022 transition and are working on plans for how to roll this out in Lambda, although at this stage we're not in a position to give a schedule. It's a little complicated because the minimum Linux kernel version is different between AL2 and AL2022.

**Comment by joebowbeer** [comment]
A general question about AWS Node.js support: What does it mean for AWS to _support_ a Node release AFTER the EOL of that release?

Using Node 16 as an example:

From [AWS Lambda adds Node.js 16](https://aws.amazon.com/about-aws/whats-new/2022/05/aws-lambda-adds-support-node-js-16/) -

> Node.js 16 is the latest long-term support (LTS) release of Node.js and will be supported for security and bug fixes until April 2024.

Given that Node 16 reaches EOL in [September, 2023](https://nodejs.org/en/about/releases/), how is AWS planning to make security and bug fixes after that date?

**Comment by rktyt** [comment]
@joebowbeer 
It seems that AWS just hasn't made any announcements regarding the shortened support period for Node.js.

AWS announcement is May 12, 2022.
Node.js Announcement is June 8, 2022 (https://nodejs.org/ru/blog/announcements/nodejs16-eol/)

----

[AWS Lambda runtime deprecation policy](https://docs.aws.amazon.com/lambda/latest/dg/lambda-runtimes.html#runtime-support-policy) shows that it can be called after EOL, but I assume it will just be left unpatched.

**Comment by jtuliani** [comment]
In general, the schedule for Lambda runtime deprecations tracks the language version they support (there have been exceptions e.g. Python 2.7). Customers should not plan on Lambda runtime support extending beyond the LTS period of the corresponding language version. This applies to the shortened EOL for Node 16.

We recognize that the shortened LTS period for Node 16 will increase demand for early support for Node 18, and are factoring this in to our Node 18 planning.

**Comment by willfarrell** [comment]
Is there a status update on when this will be ready or in preview? The next version on Middy, [v4](https://github.com/middyjs/middy/pull/908), with support for NodeJS 18, AWS SDK v3, and many more requested features is nearly ready in anticipation of an earlier release.

**Comment by jtuliani** [comment]
Work on Node 18 support for AWS Lambda is in progress. There are various dependencies which mean we're not able to share an ETA at this time.

**Comment by GeoffreyEmerson** [comment]
Node v19 released. https://nodejs.org/en/blog/release/v19.0.0/

That means v18 is feature stable, right?

**Comment by IvanMoucha** [comment]
> Node v19 released. https://nodejs.org/en/blog/release/v19.0.0/
> 
> That means v18 is feature stable, right?

The Active LTS is starting 2022-10-25. After this date, it will be feature stable.

**Comment by rohanrajpal** [comment]
Nodejs v18 is LTS now, so yep!

> Node v19 released. https://nodejs.org/en/blog/release/v19.0.0/
> 
> That means v18 is feature stable, right?

**Comment by o-alexandrov** [comment]
Oh wow, seems like AWS is rolling out Node.js 18 for AWS Lambda.
@jtuliani Could you please clarify?

<img width="1288" alt="Screenshot 2022-10-26 at 7 58 04 AM" src="https://user-images.githubusercontent.com/9992724/198032223-11df16d4-1084-4f6a-bf36-fe504573d7a1.png">

**Comment by jtuliani** [comment]
Well spotted, that was quick :)

Yes, we have shipped a Node 18 base container image, which you can use to create and deploy Node 18 functions if using container images as your function deployment format. The remainder of our Node 18 support (managed runtime, SDK, SAM CLI, CDK, AWS CLI, docs etc) will take a few more weeks, and we will make the formal announcement when all those are in place.

Note that the Node 18 image includes the AWS SDKv3 (replacing v2 in Node 16). If you're giving it a try, please bear this in mind.

**Comment by o-alexandrov** [comment]
@paul-uz Node.js 18 is currently available only for AWS Lambda functions that use container images (Docker).

> … (the rest) will take a few more weeks

**Comment by paul-uz** [comment]
> @paul-uz Node.js 18 is currently available only for AWS Lambda functions that use custom container images (Docker).
> 
> > … (the rest) will take a few more weeks

Ah lovely thank you. I wasn't sure where Cloudformation came into it you see. Thanks for clarifying.

**Comment by thdxr** [comment]
what linux kernel version is this using?

**Comment by GrahamCampbell** [comment]
the same as the node 16 ones i presume. the one al2 uses.

**Comment by jtuliani** [comment]
The kernel is not part of the runtime. All Lambda runtimes inherit their kernel from the MicroVM in which they run (see diagram on page 11 of https://docs.aws.amazon.com/pdfs/whitepapers/latest/security-overview-aws-lambda/security-overview-aws-lambda.pdf). Currently, this is kernel v4.14.

**Comment by michael-k** [comment]
Release [1.65.0](https://github.com/aws/aws-sam-cli/releases/tag/v1.65.0) of AWS SAM CLI comes with [support for Node.js 18.x](https://github.com/aws/aws-sam-cli/pull/4418).

**Comment by antspk** [comment]
[2.51.0](https://github.com/aws/aws-cdk/releases/tag/v2.51.0) AWS CDK comes with Node.js 18 runtime support as well 🤯. Keen to see when it's going to be available on codebuild as well.

**Comment by michael-k** [comment]
For CodeBuild see aws/aws-codebuild-docker-images#580

**Comment by RaphaelManke** [comment]
I've build a cdk example that uses the node 18 runtime. 
https://github.com/RaphaelManke/cdk-lambda-node-18

**Comment by o-alexandrov** [comment]
Please let us know when to expect:
- Node.js 18 for Lambda@Edge

**Comment by Sparticuz** [comment]
https://docs.aws.amazon.com/lambda/latest/dg/lambda-nodejs.html?icmpid=docs_lambda_rss

Looks like it's hit lambda proper.

**Comment by jtuliani** [comment]
Node 18 support in Lambda has shipped!

https://aws.amazon.com/about-aws/whats-new/2022/11/aws-lambda-support-node-js-18/

https://aws.amazon.com/blogs/compute/node-js-18-x-runtime-now-available-in-aws-lambda/

Thank you all for the input and feedback on this thread.

**Comment by laymonage** [comment]
Node.js 16 release: 2021-04-20
Node.js 16 active LTS start: 2021-10-26
AWS Lambda Node 16 support: 2022-05-12
LTS-to-support: 198 days

Node.js 18 release: 2022-04-19
Node.js 18 active LTS start: 2022-10-25
AWS Lambda Node 18 support: 2022-11-18
LTS-to-support: 24 days

That's a big improvement!

I remember a lot of people (myself included) complaining about how Node.js 16 support came very late. Now that Node.js 18 support came in a reasonably timely manner, I feel we need to give credit where it's due 🙂

Congratulations and thanks @jtuliani & team! 🎉 

Now, see if AWS can beat the record when Node.js 20 comes out 😜

**Comment by jlarmstrongiv** [comment]
Great news!

@jtuliani will you and the team update https://github.com/aws/aws-lambda-nodejs-runtime-interface-client to match the Node.js 18 release?  

Features described in the blog post, like ES Module support https://github.com/aws/aws-lambda-nodejs-runtime-interface-client/issues/44, do not work.  This support is very important for custom docker containers.

**Comment by paul-uz** [comment]
Excellent news, and the cherry on the top is that the ESM Module & Lamda Layer issue has also been fixed 🙏

**Comment by jtuliani** [comment]
Thank you @paul-uz and @laymonage for your kind words. We're still not where we'd like to be across other languages, but we are continuing our work to improve.

**Comment by chenrui333** [comment]
@jtuliani Congratulations on the nodejs v18 runtime release!

Worth calling out that the node.js 18 runtime seems slower than the node.js 16 runtime, any reason why that happened? Thanks!
https://twitter.com/sussmansa/status/1593804631192346624?s=46&amp;t=faTnOE-p9Hnj0qf7-KYT2g

**Comment by daniele-sarnari-blinkoo** [comment]
Do you know how to create lambda using node.js 16?
I can't create a lambda edge because nodejs 16 is no more available

**Comment by paul-uz** [comment]
> Do you know how to create lambda using node.js 16?
> I can't create a lambda edge because nodejs 16 is no more available

Why can't you use 18?

**Comment by daniele-sarnari-blinkoo** [comment]
> > Do you know how to create lambda using node.js 16?
> > I can't create a lambda edge because nodejs 16 is no more available
> 
> Why can't you use 18?

<img width="813" alt="image" src="https://user-images.githubusercontent.com/103031141/203545789-67867125-0733-43ef-a758-f73f4906efb2.png">

**Comment by daniele-sarnari-blinkoo** [comment]
nvm, fixed with
`aws lambda update-function-configuration --function-name <funcname> --runtime nodejs16.x`

**Comment by willfarrell** [comment]
Big congrats to the lambda runtime team on this release! Exciting to see the ESM + Lambda layer regression resolved. 

Middy v4 is now released for nodejs18.x runtime and now uses AWS SDK v3.

**Comment by JacksonGariety** [comment]
I'm curious about the cold start problem pointed out by @chenrui333.

**Comment by jtuliani** [comment]
@chenrui333 @JacksonGariety We're aware of the reports of longer cold starts on Node 18 and we're investigating the issue. We'll report back once those investigations are complete.

**Comment by paul-uz** [comment]
@jtuliani I found this issue, https://github.com/aws/aws-sdk-js-v3/issues/4225, when writing some new functions.

Is this intended behaviour?

I (probably wrongly) assumed that the embedded SDK would work in combination with a local node_modules folder in the function, but it seems like it doesn't.

**Comment by dvargas92495** [comment]
@jtuliani Any ETA on when node18 will be supported in Lambda@Edge?

![Screen Shot 2022-11-28 at 12 13 14 AM](https://user-images.githubusercontent.com/7143571/204199048-f1889e0d-4de2-4fce-91c7-2db32ff38fd0.png)

**Comment by aukevanleeuwen** [comment]
> @chenrui333 @JacksonGariety We're aware of the reports of longer cold starts on Node 18 and we're investigating the issue. We'll report back once those investigations are complete.

@jtuliani we are also experiencing this. Is there a GitHub issue or somewhere else where we can track the progress on this investigation? This is blocking us from moving away from 16 at the moment.

**Comment by o-alexandrov** [comment]
cc @jtuliani 
Regarding cold starts, there's a great [article by Yan Cui from a few years ago](https://miro.medium.com/max/1400/0*bkLuUfY5ypyW8rHn.webp) that is still applicable.
TLDR img from the article is attached below.

Everything in the article is still applicable.
For best performance/cold starts, bundle aws-sdk and don't depend on AWS-provided SDK
- you stop to control the version and thus, AWS could either introduce a regression you wouldn't notice until you get an error (instead of testing before deployment) or you would depend on a feature that is not released yet and works locally, but not remotely.

Unfortunately, [AWS didn't (or couldn't) listen to the community in the Node.js v16 thread](https://github.com/aws/aws-lambda-base-images/issues/14#issuecomment-1029973789), so we don't have an SDK-free node.js (hopefully soon [Bun](https://bun.sh/) also) runtime.

![coldstarts](https://user-images.githubusercontent.com/9992724/205512343-060194e0-aa6d-4f7b-a776-d2dad0b1a4ed.png)

**Comment by asilvas** [comment]
Bun support would be a perfect fit for lambda's due to fast cold start, but it's quite a few months from v1 and I doubt AWS will even consider it prior to that.

**Comment by paul-uz** [comment]
> Bun support would be a perfect fit for lambda's due to fast cold start, but it's quite a few months from v1 and I doubt AWS will even consider it prior to that.

I doubt we'll ever see the likes of Bun or even Deno. Shame.

**Comment by o-alexandrov** [comment]
@jtuliani thank you for the new feature, [AWS Lambda runtime management controls](https://aws.amazon.com/blogs/compute/introducing-aws-lambda-runtime-management-controls/).
- does it mean we could expect ability to select an `@aws-sdk`-free environment at some point?

**Comment by jtuliani** [comment]
@o-alexandrov Supporting new runtimes such as Deno or Bun would be a separate feature, it's not enabled by runtime management controls. I'm interested to learn more about your use case for Deno and/or Bun. How would they help you?

**Comment by o-alexandrov** [comment]
@jtuliani thank you for considering this request.
imho:
- `bun` would excite customers due to the performance reasons
  - community (ex. [sharp in this issue](https://github.com/lovell/sharp/issues/3511)) already considers adding support for `bun`, so I believe in such performance-critical aspects, like `image processing` `bun` would be a good fit for `AWS Lambda`
    - especially for `S3 Object Lambda`, `Lambda@Edge`, and hopefully for CloudFront's Edge functions when limits would be higher
- `deno` is less attractive to me personally than `bun`, I see `deno` being different to `node.js` only in style rather than anything measurable (including security-related differences)

**Comment by santiperone** [comment]
@jtuliani Has there been any updates on the cold start issue with node.js v18?

**Comment by paul-uz** [comment]
> @jtuliani Has there been any updates on the cold start issue with node.js v18?

What's the cold start issue?

**Comment by santiperone** [comment]
> > @jtuliani Has there been any updates on the cold start issue with node.js v18?
> 
> What's the cold start issue?

Node 18 has longer cold start in comparison to node 16

> @chenrui333 @JacksonGariety We're aware of the reports of longer cold starts on Node 18 and we're investigating the issue. We'll report back once those investigations are complete.

**Comment by jtuliani** [comment]
@santiperone Thanks for following up. We have identified the primary cause of the longer cold starts in the Node 18 runtime as thread blocking in a dependency used during initialization. We are evaluating options to resolve this in a future Node 18 runtime update.

**Comment by paul-uz** [comment]
> @santiperone Thanks for following up. We have identified the primary cause of the longer cold starts in the Node 18 runtime as thread blocking in a dependency used during initialization. We are evaluating options to resolve this in a future Node 18 runtime update.

What's the rough ETA on that update? I hope its sometime this year...

**Comment by thdxr** [comment]
that's excellent news, is it something in the aws lambda runtime interface client?

**Comment by willfarrell** [comment]
This was in a newsletter from Node Weekly I received today:
> [Lambda Cold Starts](https://maxday.github.io/lambda-perf/) is a neat page giving a live demonstration of how quickly different AWS Lambda serverless runtimes take to start. Given Node's mature status on Lambda, it does surprisingly poorly?

Seems others have noticed too.

**Comment by paul-uz** [comment]
> This was in a newsletter from Node Weekly I received today:
> > [Lambda Cold Starts](https://maxday.github.io/lambda-perf/) is a neat page giving a live demonstration of how quickly different AWS Lambda serverless runtimes take to start. Given Node's mature status on Lambda, it does surprisingly poorly?
> 
> Seems others have noticed too.

How embarrassing.

**Comment by o-alexandrov** [comment]
This is slightly related to performance questions:
- [Bun released custom runtime AWS Lambda layer](https://bun.sh/blog/bun-v0.5.7#aws-lambda-support) 

Maybe, AWS would kindly consider to provide us with managed by AWS Bun runtime.

**Comment by o-alexandrov** [comment]
> @chenrui333 @JacksonGariety We're aware of the reports of longer cold starts on Node 18 and we're investigating the issue. We'll report back once those investigations are complete.

@bicatu this ticket is closed, Node.js 18 is available for all Lambda, including Lambda@Edge (supported added in January 2023 if I'm not mistaken), some of us just shamelessly pinged AWS here 😅 having an opportunity to be heard 

I just created a separate ticket #83 regarding cold starts, so it's in `Open` state and has less probability of being postponed.

**Comment by paul-uz** [comment]
@ffxsam funnily enough, I read that same article last night.

Anyone here got a good example/tutorial on using esbuild with Typescript and tree shaking? I can't seem to find anything decent online :/

**Comment by paul-uz** [comment]
@ffxsam oh sweet. TBH I've simply not yet tried it, as we currently just use TS to compile the JS files, and bundle all the production NPM dependencies in a ZIP.

So, I can use esbuild to compile the TS into JS, at the same time, tree-shaking and minifying everything? So in theory I would end up with a JS file and a trimmed down node_modules folder? Or does esbuild create something else that gets imported?

**Comment by santiperone** [comment]
> 



> @ffxsam oh sweet. TBH I've simply not yet tried it, as we currently just use TS to compile the JS files, and bundle all the production NPM dependencies in a ZIP.
> 
> So, I can use esbuild to compile the TS into JS, at the same time, tree-shaking and minifying everything? So in theory I would end up with a JS file and a trimmed down node_modules folder? Or does esbuild create something else that gets imported?

That's right, the output from esbuild would be a single js file including all of your dependencies. If you're using SAM they already support this out of the box with not much additional config. https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/serverless-sam-cli-using-build-typescript.html.

Also note that when bundling the aws-sdk-v3 you might as well modify the MainFields property as well for better tree-shaking. This is explained here https://github.com/aws/aws-sam-cli/issues/4093.

**Comment by paul-uz** [comment]
How is that single JS then referenced in my actual function? Does esbuild handle that as well?

**Comment by simlu** [comment]
I guess this thread has now completely derailed, so I'm going to unsubscribe. However, relevant to the recent discussion, please take a look at "@vercel/ncc". We've been using it for hundreds of production lambda functions (and billions of executions) to bring down cold start / package size. It might take a bit of tinkering to get to work (depending on your dependencies - really helps if you have tests you can run against the compiled code), but the impact on cold start times is fantastic.

**Comment by paul-uz** [comment]
@ffxsam I gave esbuild a quick test, and it seems to be working, mostly. I don't think tree-shaking is working by default though, as an included package I wrote is having all its functions bundled, not just the one being used in my main file.

**Comment by koshic** [comment]
@paul-uz plz read https://esbuild.github.io/api/#tree-shaking and stop posting irrelevant comments here )
